wdl/transforms/biscayne/src/main/scala/wdl/transforms/biscayne/linking/expression/types/BiscayneTypeEvaluators.scala

```diff
@@ -161,14 +198,101 @@ object BiscayneTypeEvaluators {
+      // We don't know whether the user has provided "1" or "orange" at this stage.
+      // This is OK as-is because the value evaluators do the coercing and throw meaningful errors if the coercion fails.
+      expectedType.isCoerceableFrom(evaluatedType)
+
+    // Helper method to check something (maybe) found in the struct literal to something (maybe) found in the struct definition.
+    def checkIfMemberIsValid(typeName: String,
+                             memberName: String,
+                             evaluatedType: Option[WomType],
+                             expectedType: Option[WomType]
     ): ErrorOr[WomType] =
-      // This works fine, but is not yet a strong enough type check for the WDL 1.1 spec
-      // (i.e. users are able to instantiate struct literals with k/v pairs that aren't actually in the struct definition, without an error being thrown.)
-      // We want to add extra validation here, and return a WomCompositeType that matches the struct definition of everything is OK.
-      // Note that users are allowed to omit optional k/v pairs in their literal.
-      // This requires some extra work to be done in a subsequent PR.
-      WomObjectType.validNel
+      evaluatedType match {
+        case Some(evaluated) =>
+          expectedType match {
+            case Some(expected) =>
+              if (areTypesAssignable(evaluated, expected)) evaluated.validNel
+              else
+                s"$typeName.$memberName expected to be ${expected.friendlyName}. Found ${evaluated.friendlyName}.".invalidNel
+            case None => s"Type $typeName does not have a member called $memberName.".invalidNel
+          }
+        case None => s"Error evaluating the type of ${typeName}.${memberName}.".invalidNel
+      }
+
+    // For each member in the literal, check that it exists in the struct definition and is the expected type.
+    def checkMembersAgainstDefinition(a: StructLiteral,
+                                      structDefinition: WomCompositeType,
+                                      linkedValues: Map[UnlinkedConsumedValueHook, GeneratedValueHandle],
+                                      typeAliases: Map[String, WomType]
+    ): ErrorOr[WomCompositeType] = {
+      val checkedMembers: Map[String, ErrorOr[WomType]] = a.elements.map { case (memberKey, memberExpressionElement) =>
+        val evaluatedType =
+          expressionTypeEvaluator.evaluateType(memberExpressionElement, linkedValues, typeAliases).toOption
+        val expectedType = structDefinition.typeMap.get(memberKey)
+        (memberKey, checkIfMemberIsValid(a.structTypeName, memberKey, evaluatedType, expectedType))
+      }
+
+      val (errors, validatedTypes) = checkedMembers.partition { case (_, errorOr) =>
+        errorOr match {
+          case Invalid(_) => true
+          case Valid(_) => false
+        }
+      }
+
+      errors.headOption match {
+        case Some((_, Invalid(_))) =>
```

Comment:
Do we need the `Invalid` here? Aren't we just checking whether `errors.head` exists?

I had to read this a few times to figure out what was going on, might be easier to understand with the humble `if`. 
```
if (errors.nonEmpty)
    errors.collect ...
else
    WomCompositeType(...)
```

That might just be me, though, interested in other opinions.